src/main/java/ch/cern/eam/wshub/core/services/administration/impl/UserGroupMenuServiceImpl.java

```diff
@@ -0,0 +1,321 @@
+        MP6005_GetExtMenusHierarchy_001 getExtMenusHierarchy = new MP6005_GetExtMenusHierarchy_001();
+        getExtMenusHierarchy.setUSERGROUPID(new USERGROUPID_Type());
+        getExtMenusHierarchy.getUSERGROUPID().setUSERGROUPCODE(node.userGroup);
+
+        ExtMenusHierarchy result =
+                tools.performInforOperation(context, inforws::getExtMenusHierarchyOp, getExtMenusHierarchy)
+                        .getResultData().getExtMenusHierarchy();
+
+        return result;
+    }
+
+    private List<GenericMenuEntry> getExtMenuHierarchyAsList(InforContext context, MenuSpecification node) throws InforException {
+        ExtMenusHierarchy result = this.getExtMenuHierarchy(context, node);
+
+        List<GenericMenuEntry> menuEntries = new ArrayList<>();
+        for(MENU_Type menu : result.getMENU()) {
+            menuEntries.add(new GenericMenuEntry(menu));
+            menu.getFOLDER().stream().forEach(childFolder -> this.addFolderToEntries(menuEntries, childFolder));
+            menu.getFUNCTION().stream().forEach(childFunction -> this.addFunctionToEntries(menuEntries, childFunction));
+        }
+
+        return menuEntries;
+    }
+
+    private String decideMenuType(String menuCode, String[] path) {
+        String menuType;
+        if (menuCode == null || menuCode.isEmpty()) {
+            if (path.length <= 1) {
+                menuType = "M"; // Item is main menu
+            } else {
+                menuType = "F"; // Item is submenu
+            }
+        } else {
+            menuType = "S"; // Item is function
+        }
+
+        //WARN If two paths are identical, item will be added unreliably to one of them, as no item code is given (only path) to this function; we assume correct menu hierarchy already exists
+        return menuType;
+    }
+
+    private String[] calculateExistingPath(String[] words, List<GenericMenuEntry> menuEntries) {
+        GenericMenuEntry current = null;
+
+        int amountFound = 0;
+        for (String next : words) {
+            Boolean gmeFound = false;
+            for (GenericMenuEntry gme : menuEntries) {
+                if (!gmeFound && gme.description.equals(next) && (current == null || gme.getParent().equals(current.getId()))) {
+                    current = gme;
+                    gmeFound = true;
+                    amountFound++;
+                }
+            }
+            if (!gmeFound) {
+                return Arrays.copyOf(words, amountFound);
+            }
+        }
+
+        return words; // If reached the end, then all path could be found
+    }
+
+
+    /**
+     * Adds a full menu/submenu/function path to the menu hierarchy.
+     * @param context the user credentials
+     * @param node the specified full path and function to add, for a specific user group
+     * @return
+     */
+    @Override
+    public String addToMenuHierarchy(InforContext context, MenuSpecification node) throws InforException {
```

Comment:
This method should be separated into 2: The logic of what needs to be added and the wrapper for the infor ws invocation

Reply:
Wrapper method for infor added.